### ISSUE TITLE
Inconsistency in reporting errors on invariant arguments

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#3263]
This is related to #3258 and I'm trying to understand the logic of the error reporting.
```python
from typing import List

def test2(x: List[Hashable]) -> None:
    pass


test2(["abc", "def"])
l1 = ["abc", "def"]
test2(l1)
```

pyright v1.1.233 now reports an error on the second call to `test2()`, but not the first call.  Prior to 1.1.233, no error would be reported.

What I don't understand is why the first call is OK (no error reported), but the second call does show an error.  For the first call, wouldn't pyright determine that `["abc", "def"]` is `list[str]` and then give the error shown when `l1` is inferred to be `list[str]`?

**Comment by erictraut** [comment]
For the first call, pyright uses bidirectional inference to evaluate the type `list[Hashable]`. In the second example, the statement `l1` doesn't have any context for bidirectional inference, so pyright falls back on its default inference rules for lists, and it evaluates the type as `list[str]`. For more details, please refer to the [pyright documentation](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types).

The use of bidirectional type inference is also found in other Python type checkers (e.g. mypy and pyre) and type checkers for other languages (e.g. TypeScript).

**Comment by Dr-Irv** [comment]
OK, that makes sense, but I'm trying to figure out a way to develop tests for checking if type stubs are correct (in particular, for pandas).

Using the example above, the type of the argument `x`  as `List[Hashable]` is wrong.  But if the test only contains:
```python
test2(["abc", "def"])
```
then we think our stub is correct.  With the stubs here for pandas https://github.com/microsoft/python-type-stubs/tree/main/tests/pandas, we are mostly using that form, but are not really testing the second form:
```python
l1 = ["abc", "def"]
test2(l1)
```
So is the only way to properly test the stubs to use the second form where bidrectional type inference is not used?  Or is there some other methodology to use with respect to testing type stubs?

**Comment by erictraut** [comment]
If you want to be sure that bidirectional inference or type narrowing are not affecting the validity of your tests, you might want to use a function parameter with a declared type.

```python
def func1(l1: list[str], l2: list[Hashable]):
    test2(l1) # Generates error
    test2(l2) # Does not generate error
```
